xc/device.ml

```diff
@@ -1493,7 +1493,15 @@ module Dm = struct
 
     let qmp_event_handle domid qmp_event =
       (* This function will be extended to handle qmp events *)
-      debug "Got QMP event, domain-%d: %s" domid qmp_event.event
+      debug "Got QMP event, domain-%d: %s" domid qmp_event.event;
+      qmp_event.data >>= fun d ->
+      match d with
+      | RTC_CHANGE timeoffset ->
+        with_xs (fun xs ->
+          let timeoffset_key = "/vm/" ^ (Uuidm.to_string (Xenops_helpers.uuid_of_domid ~xs domid)) ^ "/rtc/timeoffset" in
+          let rtc = try xs.Xs.read timeoffset_key with Xs_protocol.Enoent _ -> "0" in
+          xs.Xs.write timeoffset_key (string_of_int ((int_of_string rtc) + timeoffset))
+        )
```

Comment:
It looks like xenopsd watches for the rtc/timeoffset xenstore key changes and then triggers the following code:
https://github.com/xapi-project/xenopsd/blob/master/xc/xenops_server_xen.ml#L2960
```
			| "vm" :: uuid :: "rtc" :: "timeoffset" :: [] ->
				let timeoffset = try Some (xs.Xs.read path) with _ -> None in
				Opt.iter
					(fun timeoffset ->
						(* Store the rtc/timeoffset for migrate *)
						store_rtc_timeoffset uuid timeoffset;
						(* Tell the higher-level toolstack about this too *)
						Updates.add (Dynamic.Vm uuid) internal_updates
```
I wonder if we couldn't replace the RTC_CHANGE code for something like (note that you'll need to pass a function containing :
```
somewhere in xenopsd_server_xen:
let store_rtc_and_update timeoffset =
						(* Store the rtc/timeoffset for migrate *)
						store_rtc_timeoffset uuid timeoffset;
						(* Tell the higher-level toolstack about this too *)
						Updates.add (Dynamic.Vm uuid) internal_updates

and then in device.ml:
+      | RTC_CHANGE timeoffset ->store_rtc_and_update timeoffset
```
and remove the key watching code in xenops_server_xen.ml#L2960.

The problem with this suggestion is how to make store_rtc_and_update available inside the event handler. With the xenstore key, it's easy to decouple the qmp event handler code from the xenopsd code in xenops_server_xen.ml#L2960.  @jonludlam, what do you think?